File: lib/google.ts
import { google } from 'googleapis';
import { decrypt } from './crypto';
import { one } from './db';

export function oauthClient() {
  return new google.auth.OAuth2(process.env.GOOGLE_CLIENT_ID, process.env.GOOGLE_CLIENT_SECRET, process.env.GOOGLE_REDIRECT_URI);
}

export function googleAuthUrl(state: string) {
  return oauthClient().generateAuthUrl({
    access_type: 'offline',
    prompt: 'consent',
    state,
    scope: [
      'openid', 'email', 'profile',
      'https://www.googleapis.com/auth/gmail.send',
      'https://www.googleapis.com/auth/spreadsheets'
    ]
  });
}

export async function clientForUser(userId: string) {
  const user = await one<{ google_refresh_token: string }>('SELECT google_refresh_token FROM users WHERE id=$1', [userId]);
  if (!user) throw new Error('User not found');
  const client = oauthClient();
  client.setCredentials({ refresh_token: decrypt(user.google_refresh_token) });
  return client;
}

export async function readSheet(userId: string, spreadsheetId: string, sheetName: string) {
  const auth = await clientForUser(userId);
  const sheets = google.sheets({ version: 'v4', auth });
  const res = await sheets.spreadsheets.values.get({ spreadsheetId, range: `'${sheetName.replaceAll("'", "''")}'` });
  return res.data.values ?? [];
}

export async function getSheetNames(userId: string, spreadsheetId: string) {
  const auth = await clientForUser(userId);
  const sheets = google.sheets({ version: 'v4', auth });
  const res = await sheets.spreadsheets.get({ spreadsheetId, fields: 'sheets.properties.title' });
  return (res.data.sheets ?? []).map(s => s.properties?.title).filter(Boolean) as string[];
}

export async function ensureStatusColumns(userId: string, spreadsheetId: string, sheetName: string, headers: string[]) {
  const required = ['Status', 'Sent At', 'Error'];
  const missing = required.filter(h => !headers.includes(h));
  if (!missing.length) return [...headers];
  const auth = await clientForUser(userId);
  const sheets = google.sheets({ version: 'v4', auth });
  const next = [...headers, ...missing];
  await sheets.spreadsheets.values.update({ spreadsheetId, range: `'${sheetName.replaceAll("'", "''")}'!1:1`, valueInputOption: 'RAW', requestBody: { values: [next] } });
  return next;
}

function colLetter(index: number) {
  let n = index + 1, s = '';
  while (n > 0) { const r = (n - 1) % 26; s = String.fromCharCode(65 + r) + s; n = Math.floor((n - 1) / 26); }
  return s;
}

export async function updateRecipientRow(userId: string, spreadsheetId: string, sheetName: string, rowNumber: number, headers: string[], status: string, sentAt: string, error: string) {
  const auth = await clientForUser(userId);
  const sheets = google.sheets({ version: 'v4', auth });
  const indices = ['Status', 'Sent At', 'Error'].map(h => headers.indexOf(h));
  const start = Math.min(...indices), end = Math.max(...indices);
  const vals = Array(end - start + 1).fill('');
  vals[indices[0] - start] = status;
  vals[indices[1] - start] = sentAt;
  vals[indices[2] - start] = error;
  await sheets.spreadsheets.values.update({ spreadsheetId, range: `'${sheetName.replaceAll("'", "''")}'!${colLetter(start)}${rowNumber}:${colLetter(end)}${rowNumber}`, valueInputOption: 'RAW', requestBody: { values: [vals] } });
}

export async function sendGmail(userId: string, to: string, subject: string, html: string) {
  const auth = await clientForUser(userId);
  const gmail = google.gmail({ version: 'v1', auth });
  const mime = [
    `To: ${to}`,
    `Subject: ${subject.replace(/[\r\n]+/g, ' ')}`,
    'MIME-Version: 1.0',
    'Content-Type: text/html; charset=UTF-8',
    '',
    html
  ].join('\r\n');
  const raw = Buffer.from(mime).toString('base64url');
  return gmail.users.messages.send({ userId: 'me', requestBody: { raw } });
}

File: backend/app/services/google.py
import base64
import email.utils
from email.message import EmailMessage

from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import Flow
from googleapiclient.discovery import build

from ..config import settings
from ..models import User
from ..security import decrypt

SCOPES = [
    "openid",
    "email",
    "profile",
    "https://www.googleapis.com/auth/gmail.modify",
    "https://www.googleapis.com/auth/spreadsheets",
]


def oauth_flow(
    state: str | None = None,
    code_verifier: str | None = None,
) -> Flow:
    config = {
        "web": {
            "client_id": settings.google_client_id,
            "client_secret": settings.google_client_secret,
            "auth_uri": "https://accounts.google.com/o/oauth2/auth",
            "token_uri": "https://oauth2.googleapis.com/token",
            "redirect_uris": [settings.google_redirect_uri],
        }
    }

    return Flow.from_client_config(
        config,
        scopes=SCOPES,
        state=state,
        redirect_uri=settings.google_redirect_uri,
        code_verifier=code_verifier,
        autogenerate_code_verifier=code_verifier is None,
    )

def credentials_for(user: User) -> Credentials:
    creds = Credentials(
        token=None,
        refresh_token=decrypt(user.google_refresh_token),
        token_uri="https://oauth2.googleapis.com/token",
        client_id=settings.google_client_id,
        client_secret=settings.google_client_secret,
        scopes=SCOPES,
    )
    creds.refresh(Request())
    return creds


def gmail_for(user: User):
    return build("gmail", "v1", credentials=credentials_for(user), cache_discovery=False)


def sheets_for(user: User):
    return build("sheets", "v4", credentials=credentials_for(user), cache_discovery=False)


def read_sheet(user: User, sheet_id: str, sheet_name: str) -> list[list[str]]:
    escaped_name = sheet_name.replace("'", "''")
    result = (
        sheets_for(user)
        .spreadsheets()
        .values()
        .get(spreadsheetId=sheet_id, range=f"'{escaped_name}'")
        .execute()
    )
    return result.get("values", [])


def send_message(
    user: User,
    to: str,
    subject: str,
    html_body: str,
    *,
    thread_id: str | None = None,
    in_reply_to: str | None = None,
) -> tuple[dict, str]:
    message = EmailMessage()
    message["To"] = to
    message["From"] = user.email
    message["Subject"] = subject.replace("\r", " ").replace("\n", " ")
    rfc_message_id = email.utils.make_msgid(domain="mailpilot.local")
    message["Message-ID"] = rfc_message_id
    if in_reply_to:
        message["In-Reply-To"] = in_reply_to
        message["References"] = in_reply_to
    message.set_content("This message requires an HTML-capable email client.")
    message.add_alternative(html_body, subtype="html")
    payload: dict[str, str] = {"raw": base64.urlsafe_b64encode(message.as_bytes()).decode()}
    if thread_id:
        payload["threadId"] = thread_id
    result = gmail_for(user).users().messages().send(userId="me", body=payload).execute()
    return result, rfc_message_id


def thread_has_reply(user: User, thread_id: str, recipient_email: str) -> bool:
    thread = gmail_for(user).users().threads().get(userId="me", id=thread_id, format="metadata").execute()
    recipient = recipient_email.casefold()
    for message in thread.get("messages", []):
        headers = {
            h["name"].casefold(): h["value"] for h in message.get("payload", {}).get("headers", [])
        }
        sender = headers.get("from", "").casefold()
        if recipient in sender and "SENT" not in message.get("labelIds", []):
            return True
    return False


def _column_letter(index: int) -> str:
    value = index + 1
    output = ""
    while value:
        value, remainder = divmod(value - 1, 26)
        output = chr(65 + remainder) + output
    return output


def update_sheet_status(
    user: User,
    sheet_id: str,
    sheet_name: str,
    row_number: int,
    status: str,
    last_sent_at: str = "",
    replied_at: str = "",
) -> None:
    service = sheets_for(user).spreadsheets().values()
    escaped = sheet_name.replace("'", "''")
    header_result = service.get(spreadsheetId=sheet_id, range=f"'{escaped}'!1:1").execute()
    headers = [str(value) for value in (header_result.get("values") or [[]])[0]]
    required = ["MailPilot Status", "MailPilot Last Sent", "MailPilot Replied At"]
    if any(name not in headers for name in required):
        headers.extend(name for name in required if name not in headers)
        service.update(
            spreadsheetId=sheet_id,
            range=f"'{escaped}'!1:1",
            valueInputOption="RAW",
            body={"values": [headers]},
        ).execute()
    indices = [headers.index(name) for name in required]
    start, end = min(indices), max(indices)
    values = [""] * (end - start + 1)
    values[indices[0] - start] = status
    values[indices[1] - start] = last_sent_at
    values[indices[2] - start] = replied_at
    service.update(
        spreadsheetId=sheet_id,
        range=f"'{escaped}'!{_column_letter(start)}{row_number}:{_column_letter(end)}{row_number}",
        valueInputOption="RAW",
        body={"values": [values]},
    ).execute()
